source/tutorials/source-file-selection.md

```diff
@@ -0,0 +1,241 @@
+(source-file-selection)=
+# Source file selection
+<!-- Note on title choice: While there's more uses outside of sources, it's by far the most prominent one -->
+
+To build a local project in a Nix derivation, its source files must be accessible to the builder.
+But since the builder runs in an isolated environment (if the [sandbox](https://nixos.org/manual/nix/stable/command-ref/conf-file.html#conf-sandbox) is enabled),
+it won't have access to the local project files by default.
+
+To make this work regardless, the Nix language has certain builtin features to copy local paths to the Nix store,
+whose paths are then accessible to derivation builders [^1].
+
+[^1]: Technically only Nix store paths from the derivations inputs can be accessed,
+but in practice this distinction is not important.
+
+These builtin features are very limited in functionality and are not recommended for anything non-trivial. For more advanced use cases, the file set library should be used instead.
+
+In this tutorial you'll learn both how to use the builtins and the file set library.
+
+## Setting up a local experiment
+
+To experiment with source file selection, we'll set up a local project.
+
+To start out, create a new directory, enter it, and set up `niv` to manage the Nixpkgs dependency:
```

Comment:
Can we use flakes instead of `niv`? My reasons behind this request:

0) There is a non-trivial cost to the proliferation of certain "third party" tools: https://ayats.org/blog/no-flake-utils/

1) It is already hard to navigate the Nix ecosystem (in terms of learning) for a newcomer, due to how long it has taken flakes to just be *the* way to do things. Now that community consensus has been built, official tutorials should focus on using flakes rather than alternatives to flakes? If we would like to still continue to avoid using flakes, then we ought to use a Nix-only solution?

3) It will be helpful to have more tutorials showing flake usage in various scenarios, without involving `flake-utils`, `flake-compat`, etc. (as most other blog tutorials tend to do).

Reply:
Some time ago we made the decision to generally avoid documenting experimental features in official documentation. The documentation team is very constrained on resources, and we don't want to spend this on documenting features that have the potential to break at any time. Especially when there's also an abundance of stable features that need to be documented.

As the new CLI and Flakes [get stabilised](https://discourse.nixos.org/t/stabilising-the-new-nix-command-line-interface/35531), the documentation team will definitely be on-board to update the documentation to use it though. So, I'd like to redirect any complaints about Flakes not being documented to the Nix team, which notably also doesn't have a lot of resources, they need help with stabilisation! :smile: 

Note that I am writing this tutorial fairly generically though, using a `package.nix` file that works regardless of Flakes, and the `default.nix` file can also be imported from Flakes.

It's unfortunate that there's no stable _official_ tool for locking dependencies, but I think `niv` is fine because it only relies on stable features and is generally considered stable itself.

Nevertheless, I will have to add some more notes on Flakes into this tutorial, but not in a good way: Due to Flakes always importing the local directory into the Nix store, it's a dangerous (secret files can easily get added to the store) and wasteful (many store imports for the same directory even if barely anything changed). :sweat_smile: